Add Policy Check Handler to check if packages contain prettier

jikim-msft opened the pull request:
This PR adds a policy-check handler which checks that prettier exists in every `package.json`

jikim-msft committed:
add handler npm-package-json-prettier

jikim-msft committed:
Empty-Commit

jikim-msft committed:
remove change

jikim-msft committed:
revert change

jikim-msft committed:
JSON parse error resolve

jikim-msft committed:
update pnpm-lock.yaml

jikim-msft committed:
set true as default val

jikim-msft committed:
check extraneous file

jikim-msft committed:
revert back to initial

jikim-msft committed:
fix lint

Josmithr commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,6 +585,30 @@ export const handlers: Handler[] = [
+
+            if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty)) {
+                return "Error parsing JSON file: " + file;
+            }
```
We should explicitly return `undefined` at the end of this function for the success case.

jikim-msft replied:
Added `return undefined` outside of the `if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty))` clause.

Josmithr reviewed (changes requested):
Looking good! My main feedback is about improving our error messaging, and adding a `resolver` to automatically "fix" the missing script issues.

jikim-msft committed:
change error statement & return value in success case

jikim-msft commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,6 +585,30 @@ export const handlers: Handler[] = [
+            const hasFormatProperty = Object.prototype.hasOwnProperty.call(json, "format");
+
+            if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty)) {
+                return "Error parsing JSON file: " + file;
```
Return statement changed to return the `file` and list of missing scripts. Missing scripts are pushed to `missingScripts` array if it contains `false` value.

Josmithr commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,6 +585,47 @@ export const handlers: Handler[] = [
+                //return "Error parsing JSON file: " + file;
+                const missingScripts: string[] = [];
+
+                if (hasPrettierProperty === false) {
```
Since `hasPrettierProperty` is a boolean, this can be cleaned up as:
```suggestion
                if (!hasPrettierProperty) {
```
Unless I'm missing something.

jikim-msft committed:
remove comment

jikim-msft committed:
improve logic

Josmithr commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,6 +585,47 @@ export const handlers: Handler[] = [
+                    missingScripts.push(`format`)
+                }
+
+                return file + ` missing following scripts in package.json` + missingScripts.join("\n");
```
Nit:
```suggestion
                return `${file} is missing the following scripts: ${missingScripts.join("\n\t")}`;
```

We generally recommend using the string interpolation syntax where possible. This file isn't doing a great job at that, but we might as well us it with new code.

I'm also suggesting adding a tab before each newline entry to make the formatting a bit nicer, but feel free to ignore / do what you think looks best :)

jikim-msft replied:
Ah, I missed using string interpolation. Thank you for letting me know 👍

jikim-msft committed:
Update build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts

Co-authored-by: Joshua Smithrud <54606601+Josmithr@users.noreply.github.com>

tylerbutler commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,6 +585,46 @@ export const handlers: Handler[] = [
+            );
+            const hasFormatProperty = Object.prototype.hasOwnProperty.call(json, "format");
+
+            if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty)) {
```
I think you can remove this if-check since you're now checking them individually.

jikim-msft replied:
Then would I have to return error if the length of `missingScripts` is greater or equal to 1?

tylerbutler replied:
>0, I think.

tylerbutler replied:
Oh, that's the same, I guess. Ignore me. 🤦

jikim-msft committed:
add resolver logic

jikim-msft committed:
change logic

jikim-msft commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
+                return "Error parsing JSON file: " + file;
+            }
+
+            const hasPrettierProperty = Object.prototype.hasOwnProperty.call(json, "prettier");
```
Change applied.

jikim-msft commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
+                return "Error parsing JSON file: " + file;
+            }
+
+            const hasPrettierProperty = Object.prototype.hasOwnProperty.call(json, "prettier");
```
I think we should be calling the checks on `json.scripts`, since the `handler` is checking whether the specific scripts exist or not.

Josmithr replied:
We should probably update all of the script checks to look under `json.scripts` then too, even though they are less likely to have conflicts under "dependencies", etc.

jikim-msft commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
+    if (jsonDevDependencies === undefined || jsonScripts === undefined) {
+        // create deveDependencies and add prettier
+        if (jsonDevDependencies === undefined) {
+            json["devDependencies"] = { prettier: "~2.6.2" };
```
Changed to `"prettier": undefined`

jikim-msft replied:
Changed to `"prettier" : ""`

jikim-msft committed:
improve logic

jikim-msft committed:
fix logic ii

jikim-msft commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
+        if (jsonScripts === undefined) {
+            json["scripts"] = {
+                "format": "npm run prettier:fix",
+                "prettier": "prettier --check . --ignore-path ../../../.prettierignore",
```
Change applied to discard after `--check .`

jikim-msft committed:
refactor

tylerbutler commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
+    }
+
+    // package.json in root of the release-group should have lerna script
+    if (jsonCheckRoot === "root") {
```
No, this isn't a good way. That said, I don't know of a better one outside of the release tools context, which is way too heavy to load in a policy handler. Need to think about it some more...

jikim-msft committed:
refactor ii

alexvy86 commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +586,102 @@ export const handlers: Handler[] = [
+        jsonScripts["format"] = "lerna run format --no-sort --stream -- -- -- --color";
+    }
+
+    return undefined;
```
Seems like this function doesn't need to return.

jikim-msft replied:
Removed `return undefined;`

alexvy86 commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +586,102 @@ export const handlers: Handler[] = [
+                return "Error parsing JSON file: " + file;
+            }
+
+            const hasPrettierScript = Object.prototype.hasOwnProperty.call(
+                json.scripts,
+                "prettier",
+            );
+            const hasPrettierFixScript = Object.prototype.hasOwnProperty.call(
+                json.scripts,
+                "prettier:fix",
+            );
+            const hasFormatScript = Object.prototype.hasOwnProperty.call(json.scripts, "format");
```
I think the error you're seeing (`TypeError: Cannot convert undefined or null to object`) is coming from here, when analyzing a file that does not have a `scripts` section.

<img width="378" alt="image" src="https://user-images.githubusercontent.com/716334/204929158-e49b6ed3-cd7c-4848-9454-6a302972568d.png">

Do we need to do the check with that API for some reason, instead of `json.scripts?["prettier"]`?

jikim-msft replied:
As @tylerbutler mentioned below, using it for the purpose of lint rule.

alexvy86 commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +586,102 @@ export const handlers: Handler[] = [
+            );
+            const hasFormatScript = Object.prototype.hasOwnProperty.call(json.scripts, "format");
+
+            if (!(hasPrettierScript && hasPrettierFixScript && hasFormatScript)) {
```
I think this would be more readable without the outer `if`, and after the three individual checks something like `return missingScripts.length > 0 ? <the error> : undefined;`

jikim-msft replied:
Removed the outermost if-condition and replaced with ternary operator.

alexvy86 commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +586,102 @@ export const handlers: Handler[] = [
 
+function addPrettier(json: Record<string, any>) {
+    const jsonCheckRoot = json.name;
+    let jsonDevDependencies = json.devDependencies;
+    let jsonScripts = json.scripts;
```
Do we need the two local variables? Feels like the rest of the function should be able to check/update `json["property-name"]` in place.

jikim-msft replied:
Both local variables removed.

tylerbutler reviewed (commented):
Some high-level feedback about the design and intent of this policy handler.

TL;DR: I think you can simplify quite a bit and be _less_ smart, and in this case I think that would be better.

1. I recommend making the resolver conservative. Don't just blast the scripts everywhere. Instead, maybe something like, "If the package has one of the prettier, format, or prettier:fix scripts, then all scripts should be present." The reason for this suggestion is that our repo is quite heterogeneous, and policy-check doesn't have many per-project configuration options. Thus, if the resolver is heavy-handed, you're more likely to "accidentally" affect projects you might not want to. Even if you think it should apply to every project, I predict that you'll find a few that you don't.
2. A corollary to 1: if a package has no scripts section, I wouldn't force add it. Essentially I am arguing for an "opt-in" model for enforcement. There doesn't seem to be an opt-out approach.
3. I wouldn't worry about checking the release group root package.jsons. There are very few release groups, and there's no good way to differentiate at the policy level between root packages and non-root. If we need such a thing, we should add it. (We could use the presence of a `fluidBuild` node as an indicator but build-tools doesn't have one right now.)

jikim-msft committed:
refactor iii

alexvy86 commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +586,93 @@ export const handlers: Handler[] = [
 
+function addPrettier(json: Record<string, any>) {
+    // if the json file contains scripts field
+    console.log(1);
```
[![image](https://user-images.githubusercontent.com/716334/206280064-d0a31c31-9f5f-4d35-a46c-42c1a145f0b1.png)
](https://media3.giphy.com/media/8abAbOrQ9rvLG/giphy.gif?cid=ecf05e47b8ezjn4qjpuei8gf3uq49gh8rlsvs2yu9yhk85jz&rid=giphy.gif&ct=g)

alexvy86 commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +586,93 @@ export const handlers: Handler[] = [
+    // if the json file contains scripts field
+    console.log(1);
+
+    const hasScriptsField = Object.prototype.hasOwnProperty.call(json, "scripts");
+
+    if (hasScriptsField) {
```
I think this check is worth extracting out to the `resolver` above, so the `addPrettier` function always does exactly what its name suggests, instead of sometimes not doing anything. That would also save us the `writeFile()` call on line 639 which only needs to happen if the file was actually changed.

alexvy86 commented on build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:
```diff
@@ -585,8 +586,93 @@ export const handlers: Handler[] = [
+    const hasScriptsField = Object.prototype.hasOwnProperty.call(json, "scripts");
+
+    if (hasScriptsField) {
+        const hasPrettierScriptResolver = Object.prototype.hasOwnProperty.call(
```
Is there an advantage to this way of doing the call over `json.scripts.hasOwnProperty("prettier")`? Or something like suggested in [this comment](https://github.com/microsoft/FluidFramework/pull/13072/files#r1036546558).

alexvy86 replied:
I see. I think I like `json.scripts?.prettier !== undefined` better anyway 😄 .

And aren't you supposed to be OOF @tylerbutler ? Shoo! 😛

jikim-msft committed:
refactor logic, add script check to handler, trigger helper iff script-field exists

jikim-msft commented:
Made changes according to @alexvy86 @tylerbutler 's latest feedback. 

1. `Handler`
- Added `hasScriptsField` to check if each `package.json` contains `scripts` field. Initially, this check was not included and created `TypeError: Cannot convert undefined or null to object`
- If `missingScripts` has length greater or equal to 1, it will raise a warning, otherwise return `undefined`. 

2.`Resolver`
- Moved `hasScriptsField` from the helper function `addPrettier`.  With the modified clause, `addPrettier` and `writeFile` will be invoked if and only if the `package.json` contains `scripts` field. 

3. `AddPrettier`
- As @tylerbutler 's suggestion, `addPrettier` function takes a more conservative approach in terms of adding the scripts. First, it checks whether the `package.json` contains `format`, `prettier`, and `prettier:fix` script. If it has at least one script inside it, it will decide to include all three scripts. If not, this step will be ignored. 


I believe logic in `handler` and `resolver` are somewhat repetitive (since they are both initializing same boolean variable `hasScriptsField`). However, due to the way `handler` and `resolver` are separately structured, I believe `scripts` field check in both parts are inevitable. If there seems to be a better way, I would love any suggestions. 

In order to test this locally, run the following commands in:
1. `build-tools/packages/build-tools`: `npm run build`
2. `root`: `build-tools/packages/build-cli/bin/dev check policy --exclusions build-tools/packages/build-tools/data/exclusions.json --fix -D assert-short-codes`

alexvy86 reviewed (approved):
Seems ok to me that handler and resolver do their check for scripts separately.

jikim-msft commented:
> Seems ok to me that handler and resolver do their check for scripts separately.

Great! Thanks for reviewing the PR!

Josmithr reviewed (approved):
Looks good to me!
